### PR TITLE
REV-1338 Add shadow call to  SDNFallback

### PR DESCRIPTION
**Pull request by dianekaplan**
**Background:** 
Before using our new SDN csv fallback check as the fallback for when the SDN API is down, first we'd like to gather data on how consistently the two behave. Ticket REV-1338 is to add a 'shadow' call of checkSDNFallback and log the results for both, including the inputted name/city/country for investigation in the case that the SDN API returns a hit but our fallback does not (third case below). 

I've also added handling for the case where there is no fallback data populated (true for engineer local environments until they run the management command) so we log a message of throwing an exception. 

**Example output messages:** 
SDNFallback compare: MATCH. Results - SDN API: 0 hit(s); SDN Fallback match: False. Basket: 999
SDNFallback compare: MATCH. Results - SDN API: 1 hit(s); SDN Fallback match: True. Basket: 999
SDNFallback compare: MISMATCH. Results - SDN API: 0 hit(s); SDN Fallback match: True. Basket: 999
SDNFallback compare: MISMATCH. Results - SDN API: 1 hit(s); SDN Fallback match: False. Basket: 999

And additionally for that last problem case: 
Failed SDN match for first name: Sheikh Yusuf, last name: AASI, city: Beirut, country: LB

9/25 update: added improved exception handling based on feedback
9/29 update: refactored to clean up try/except/else
9/30 update: refactored to move comparison code into sdn.py instead of cybersource.py (so the code/tests can live together in the right place) and added tests for the logged output

Relevant ticket: https://openedx.atlassian.net/browse/REV-1338

### CODE REVIEW

File: ecommerce/extensions/payment/views/cybersource.py [hunk 1]
```diff
@@ -84,21 +84,41 @@ def check_sdn(self, request, data):
...
                 hit_count,
+                SDNFallback_hit,
+                result_string
             )
             response_to_return = {
                 'error': 'There was an error submitting the basket',
                 'sdn_check_failure': {'hit_count': hit_count}}
 
+            if not SDN_calls_matched:  # Log troubleshooting info if API gets hit and fallback doesn't
+                logger.info('Failed match for first name: %s  , last name: %s , city: %s, country: %s ',
```
Comment (julianajlk) [inline ecommerce/extensions/payment/views/cybersource.py]: Can you add SDN? "Failed SDN match..." or something similar. Or even include part of your comment that the SDN API got a hit, fallback did not, so it's more explicit that this is the case of API got a bad user, and our fallback got a good user.

**Comment by jhan217** [comment]
Is there a way of structuring this so that the new, temporary functionality is completely decoupled from the existing SDN check code?

File: ecommerce/extensions/payment/views/cybersource.py (hunk 1 above)
Comment (dianekaplan) [reply ecommerce/extensions/payment/views/cybersource.py]: Sure! Right now it's the line immediately after, like so: 
```
- SDNCheck function called for basket [102]. It received 1 hit(s). SDNFallback call got: True. SDNFallback MATCH
- Failed match for: first name: Sheikh Yusuf last name: AASI , city: Beirut, country: LB
```

But I can update it to "Failed SDN match"

Comment (dianekaplan) [reply ecommerce/extensions/payment/views/cybersource.py]: Updated now- thanks!

**Comment by dianekaplan** [comment]
> Is there a way of structuring this so that the new, temporary functionality is completely decoupled from the existing SDN check code?

Done! Having it separately will mean redundant logging with what's already there, but the big upside is it'll be way simpler to remove this code after the analysis is done.

**Comment by jhan217** [comment]
> That would certainly make it easier to clean up. The benefit this way is having the logging/comparison together so Seth doesn't need to cross-reference multiple log lines in splunk for the same basket. I'll think about whether there's a way to preserve that and make cleanup a little easier later...

In addition to an easier cleanup, a couple of other reasons to decouple the code: 
1. Modifying the existing code could introduce subtle/accidental bugs into the checkout process, which may/may not be easily detected, and definitely not something we'd want to do in this scenario if we can avoid it
2. Easier to write the unit test - you wouldn't need to mock out as many things to set up the correct state + you'd only be testing the net new functionality (and none of the old existing stuff)
3. Logging - since the code is supposed to be temporary, introducing additional duplicate logs shouldn't be too big of a concern, as long as it's easy to find + separate them from the existing log. 
Edited to add: 4. Readability - it's easier for someone unfamiliar with the code to navigate and understand what the code is doing if the code is split up into separate logical chunks, as opposed to one giant long function that's doing a lot of things.

One potential way of doing this: 
```
def check_sdn(self, request, data):
 # existing code
 try:
    call_my_function(data_from_actual_call):
except Exception:
   log("something broke in the comparison but we don't care as much since the live sdn API check was 100% not affected")

def call_my_function(data_from_actual_call):
  try:
     data_from_sdn_fallback = checkSDNFallback(..)
     if data_from_actual_call != data_from_sdn_fallback:
        additional_pii = need_additional_data ? get_pii : ""
        log("SDNFallback Comparison Results - Doesn't Match, additional info: %s", additional_pii)
     else
        log("SDNFallback Comparison Results - Matches!")
  except some_specific_exception_from_check_sdn_fallback:
     log("SDNFallbackMetadata matching query does not exist.")
```

File: ecommerce/extensions/payment/core/sdn.py [hunk 2]
```diff
@@ -79,16 +79,24 @@ def checkSDNFallback(name, city, country):
...
+        )
+        records = records.filter(countries__contains=country)
+        processed_name, processed_city = process_text(name), process_text(city)
+        for record in records:
+            record_names, record_addresses = set(record.names.split()), set(record.addresses.split())
+            if (processed_name.issubset(record_names) and processed_city.issubset(record_addresses)):
+                return True
+        return False
+
+    except Exception as e:  # pylint: disable=broad-except
```
Comment (jhan217) [inline ecommerce/extensions/payment/core/sdn.py]: Why are we catching all exceptions here?

File: ecommerce/extensions/payment/views/cybersource.py [hunk 3]
```diff
@@ -70,6 +70,32 @@ class CybersourceOrderInitiationView:
     A baseclass that includes pre-work before submitting an order to cybersource for
     payment validation.
     """
+    # Temporary checkSDNFallback comparison call (REV-1338)
+    def compare_SDNCheck_vs_fallback(self, hit_count, request, data):
```
Comment (jhan217) [inline ecommerce/extensions/payment/views/cybersource.py]: Should we document the method with a description, the expected inputs and outputs?

File: ecommerce/extensions/payment/views/cybersource.py [hunk 4]
```diff
@@ -70,6 +70,32 @@ class CybersourceOrderInitiationView:
...
+        try:
+            SDNFallback_hit = checkSDNFallback(
+                data['first_name'] + ' ' + data['last_name'],
+                data['city'],
+                data['country'])
+
+            SDN_calls_matched = (hit_count > 0 and SDNFallback_hit) or (hit_count == 0 and not SDNFallback_hit)
+            result_string = "SDNFallback MATCH" if SDN_calls_matched else "SDNFallback MISMATCH"
+            logger.info(
+                'SDNCheck function called for basket [%d]. It received %d hit(s). SDNFallback call got: %s. %s',
```
Comment (jhan217) [inline ecommerce/extensions/payment/views/cybersource.py]: Could this log message more clearly delineate the SDN API check results vs the SDN fallback check results? Also, should the most important information be put closer to the beginning the the message so it'll be easier to search the logs later?

File: ecommerce/extensions/payment/views/cybersource.py [hunk 5]
```diff
@@ -70,6 +70,32 @@ class CybersourceOrderInitiationView:
...
+                request.basket.id,
+                hit_count,
+                SDNFallback_hit,
+                result_string
+                )
+
+            if hit_count > 0 and not SDN_calls_matched:
+                logger.info('Failed SDN match for first name: %s  , last name: %s , city: %s, country: %s ',
```
Comment (jhan217) [inline ecommerce/extensions/payment/views/cybersource.py]: If we wanted to quickly get all of the results of the SDN check comparisons in one query, how would we go about doing that?

File: ecommerce/extensions/payment/views/cybersource.py [hunk 6]
```diff
@@ -84,6 +110,7 @@ def check_sdn(self, request, data):
             data['city'],
             data['country'])
 
+        self.compare_SDNCheck_vs_fallback(hit_count, request, data)  # Temporary comparison call (REV-1338)
```
Comment (jhan217) [inline ecommerce/extensions/payment/views/cybersource.py]: 🎱 not sure we need the additional comment since that's already noted in the method description

File: ecommerce/extensions/payment/core/sdn.py [hunk 7]
```diff
@@ -79,16 +79,24 @@ def checkSDNFallback(name, city, country):
...
+        error_string = str(e)
+        logger.warning("SDNFallback Exception occurred: [%s]. ", e)
+        if error_string == "SDNFallbackMetadata matching query does not exist.":
+            logger.warning("SDNFallbackMetadata was empty, you probably need to run the populate_sdn_fallback_data_and_metadata.py management command")  # pylint: disable=line-too-long
```
Comment (jhan217) [inline ecommerce/extensions/payment/core/sdn.py]: Can we fix the issue instead of disabling the pylint?

Comment (dianekaplan) [reply ecommerce/extensions/payment/core/sdn.py]: I started with a shorter message, but Matthew pointed out that including the specifics of what's happened and what specific command needs to be run will be most helpful for the developer seeing the message.  I think that's a good point, and worth making the 'line too long' for pylint

Comment (jhan217) [reply ecommerce/extensions/payment/core/sdn.py]: Maybe something like this?

```
logger.warning(
   "The SDNFallbackMetadata table is empty, run `python populate_sdn_fallback_data_and_metadata.py`"
)
```

File: ecommerce/extensions/payment/views/cybersource.py (hunk 4 above)
Comment (dianekaplan) [reply ecommerce/extensions/payment/views/cybersource.py]: Moved match/mismatch bit to the front now- that is clearer you're right: 
SDNFallback compare: MATCH: SDNCheck function called for basket [106]. It received 1 hit(s). SDNFallback call got: True.

Not sure what you mean by delineating the two results?

File: ecommerce/extensions/payment/core/sdn.py (hunk 7 above)
Comment (jhan217) [reply ecommerce/extensions/payment/core/sdn.py]: +/- string concatenation to break it up if needed

🎱 Might change it to use the full tablename `payment_sdnfallbackmetadata`, in case anyone wanted to manually populate the table to circumvent running the script

File: ecommerce/extensions/payment/core/sdn.py (hunk 2 above)
Comment (dianekaplan) [reply ecommerce/extensions/payment/core/sdn.py]: The exception that had been thrown when this data isn't populated didn't seem to have a specific type to look for- not sure if there's a good way to narrow it down more if it doesn't specify? (You can see it locally if you try to make a purchase and *don't* run the management command to populate the data)

File: ecommerce/extensions/payment/views/cybersource.py (hunk 4 above)
Comment (jhan217) [reply ecommerce/extensions/payment/views/cybersource.py]: :8ball: Something like this so it makes more sense to people who don't have as much context on what a "SDNCheck function" or a "SDNFallback call" are (e.g. Seth)?
 
`SDN Comparison: MATCHED. Results - SDN API: 5 match(es); SDN Fallback: 0 match(es).`

File: ecommerce/extensions/payment/core/sdn.py (hunk 2 above)
Comment (jhan217) [reply ecommerce/extensions/payment/core/sdn.py]: Could this specific exception be caught/handled as a custom exception? Adding error handling for this case could potentially be a separate PR since it's not directly related to the shadowing task (and only seems to affect local development).

File: ecommerce/extensions/payment/core/sdn.py (hunk 7 above)
Comment (dianekaplan) [reply ecommerce/extensions/payment/core/sdn.py]: I think I'll just leave it with the script info since that's consistent with production/stage- and broke into multiple lines so it's no longer too long

File: ecommerce/extensions/payment/core/sdn.py (hunk 2 above)
Comment (dianekaplan) [reply ecommerce/extensions/payment/core/sdn.py]: Looking at custom exceptions now- yes that would be better! This should be a quick update to *make*, but not as simple to test now that my data is populated. Will leave the primitive exception handling to start since the introduction of the shadow call would otherwise break purchases for (a) local dev where management command wasn't run and (b) openedX instances (won't have the management command), and will look to come back and make the custom exception change as time allows.

File: ecommerce/extensions/payment/views/cybersource.py (hunk 3 above)
Comment (dianekaplan) [reply ecommerce/extensions/payment/views/cybersource.py]: Updated!

File: ecommerce/extensions/payment/views/cybersource.py (hunk 6 above)
Comment (dianekaplan) [reply ecommerce/extensions/payment/views/cybersource.py]: removed

File: ecommerce/extensions/payment/views/cybersource.py (hunk 5 above)
Comment (dianekaplan) [reply ecommerce/extensions/payment/views/cybersource.py]: Now that result is moved to the front, we can search for: "SDNFallback compare: "

File: ecommerce/extensions/payment/core/sdn.py (hunk 2 above)
Comment (jhan217) [reply ecommerce/extensions/payment/core/sdn.py]: Can you note the decision in the code and put in an `else raise` clause so we don't ignore unknown exceptions?

Comment (dianekaplan) [reply ecommerce/extensions/payment/core/sdn.py:94]: We actually don't want to raise any exceptions from this shadow call (and potentially block a transaction from happening)- instead I'll make sure any exceptions are logged and we'll make sure that the analysis includes checking for those logs. Then when it comes time to actually *use* the fallback and we've addressed any surprises we see, we'll raise exceptions.

File: ecommerce/extensions/payment/views/cybersource.py [hunk 8]
```diff
@@ -84,28 +84,41 @@ def check_sdn(self, request, data):
...
                 hit_count,
-                SDNFallback_result
+                SDNFallback_hit,
+                result_string
             )
             response_to_return = {
                 'error': 'There was an error submitting the basket',
                 'sdn_check_failure': {'hit_count': hit_count}}
 
+            if not SDN_calls_matched:  # Log troubleshooting info if API gets hit and fallback doesn't
```
Comment (julianajlk) [inline ecommerce/extensions/payment/views/cybersource.py]: Is there a log for if the SDNFallback gets a hit and the API doesn't?

Comment (dianekaplan) [reply ecommerce/extensions/payment/views/cybersource.py]: Yes!  We'll log: "SDNFallback compare: MISMATCH: SDNCheck function called for basket [101]. It did not receive a hit. SDNFallback call got: True. " (I've added this into the examples list in the description now for completeness). As per our parking lot chat, this one will just say that it was a mismatch and which call did/didn't hit, but no PII details, that's only for the other mismatch case.

File: ecommerce/extensions/payment/core/sdn.py (hunk 2 above)
Comment (dianekaplan) [reply ecommerce/extensions/payment/core/sdn.py]: yes- now that the rest of the items are resolved, I'm working on an updated version of the branch that will 'else raise' other kinds of exceptions from checkSDNFallback, but catch it when we call it from 'compare_SDNCheck_vs_fallback', so nothing will be raised by doing this shadow check (so we don't interfere with purchases) but so we have the option of bubbling it up down the road when it's called as an actual fallback.

File: ecommerce/extensions/payment/core/sdn.py [hunk 9]
```diff
@@ -78,17 +79,31 @@ def checkSDNFallback(name, city, country):
...
+        processed_name, processed_city = process_text(name), process_text(city)
+        for record in records:
+            record_names, record_addresses = set(record.names.split()), set(record.addresses.split())
+            if (processed_name.issubset(record_names) and processed_city.issubset(record_addresses)):
+                return True
+        return False
+
+    except Exception as e:  # pylint: disable=broad-except
+        error_string = str(e)
+        logger.warning("SDNFallback Exception occurred: [%s]. ", e)
```
Comment (jhan217) [inline ecommerce/extensions/payment/core/sdn.py]: 🎱 Move this into the else clause? Otherwise we'll get two logger warnings for the empty metadata and none for any other errors.

File: ecommerce/extensions/payment/core/sdn.py [hunk 10]
```diff
@@ -78,17 +79,31 @@ def checkSDNFallback(name, city, country):
...
+        if error_string == "SDNFallbackMetadata matching query does not exist.":
```
Comment (jhan217) [inline ecommerce/extensions/payment/core/sdn.py]: Should the function responsible for causing this error (`get_current_records_and_filter_by_source_and_type` as far as I can tell) be the one throwing the custom `SDNFallbackDataEmptyError`? That way, you don't need to do any additional error handling here and the error is easily traceable to the function that produced it. It also means that if any other functions were to call `get_current_records_and_filter_by_source_and_type`, they won't have to duplicate the same error handling code.

File: ecommerce/extensions/payment/views/cybersource.py [hunk 11]
```diff
@@ -70,6 +70,38 @@ class CybersourceOrderInitiationView:
     A baseclass that includes pre-work before submitting an order to cybersource for
     payment validation.
     """
+    def compare_SDNCheck_vs_fallback(self, request, data, hit_count):
+        """
+        Temporary checkSDNFallback comparison call (REV-1338)
```
Comment (jhan217) [inline ecommerce/extensions/payment/views/cybersource.py]: Do we have a standard format for function comments? I do like the style in https://github.com/edx/ecommerce/pull/3176/files#diff-b468567a69d0951146be670a3d37c24aR166-R179, since it's clear where the description begins/ends + describes what the function is doing in plain English, what the input parameters are (object types, business value of each input), and what the return value is.

File: ecommerce/extensions/payment/views/cybersource.py [hunk 12]
```diff
@@ -70,6 +70,38 @@ class CybersourceOrderInitiationView:
...
+            )
+
+            if hit_count > 0 and not SDN_calls_matched:
+                logger.info('Failed SDN match for first name: %s, last name: %s, city: %s, country: %s ',
+                            data['first_name'],
+                            data['last_name'],
+                            data['city'],
+                            data['country'])
+        except Exception:   # pylint: disable=broad-except
+            logger.warning(
```
Comment (jhan217) [inline ecommerce/extensions/payment/views/cybersource.py]: Where will the stacktrace be if someone wanted to dig further into this warning message to see what caused the issue?

**Review by jhan217 — COMMENTED** [review]
Looks alright, couple of comments. Needs some unit test coverage

File: ecommerce/extensions/payment/core/sdn.py (hunk 10 above)
Comment (dianekaplan) [reply ecommerce/extensions/payment/core/sdn.py]: That's a good point- will look at moving this today

File: ecommerce/extensions/payment/views/cybersource.py (hunk 11 above)
Comment (dianekaplan) [reply ecommerce/extensions/payment/views/cybersource.py]: Could you confirm which file/lines? The link above seems to re-load the PR and I'm not seeing any section with lines 166-179

File: ecommerce/extensions/payment/core/sdn.py (hunk 10 above)
Comment (dianekaplan) [reply ecommerce/extensions/payment/core/sdn.py]: done!

File: ecommerce/extensions/payment/views/cybersource.py (hunk 12 above)
Comment (dianekaplan) [reply ecommerce/extensions/payment/views/cybersource.py]: Looking earlier in the logs, they'll see the message about running the management command if they got the new SDNFallbackDataEmptyError exception, otherwise if it was something else that exception will be raised from SDNFallbackData.get_current_records_and_filter_by_source_and_type and they'd see the error info there. Is that what you mean?

Comment (jhan217) [reply ecommerce/extensions/payment/views/cybersource.py]: If that's the case, do we need this logger here, if the errors are already being logged elsewhere? What value does logging this warning here bring?

File: ecommerce/extensions/payment/core/sdn.py (hunk 9 above)
Comment (dianekaplan) [reply ecommerce/extensions/payment/core/sdn.py]: Had a great convo with Emma today about the else clause :) Did some cleanup now to separate out just the needed line in the try, and moving the other stuff down into the else.

File: ecommerce/extensions/payment/views/cybersource.py (hunk 12 above)
Comment (dianekaplan) [reply ecommerce/extensions/payment/views/cybersource.py]: I wanted to have a common line to look for in splunk to see if there were any issues preventing some purchases from doing the comparison, since we have one 'known' error path, and then others could be unknown- this way we can search splunk "Skipped SDN API vs fallback comparison; exception calling checkSDNFallback" to see any of them. Is there a better way to go about this?

File: ecommerce/extensions/payment/views/cybersource.py (hunk 11 above)
Comment (jhan217) [reply ecommerce/extensions/payment/views/cybersource.py]: https://github.com/edx/ecommerce/blob/master/ecommerce/extensions/payment/core/sdn.py#L258-L265

File: ecommerce/extensions/payment/views/cybersource.py (hunk 12 above)
Comment (jhan217) [reply ecommerce/extensions/payment/views/cybersource.py]: I think the original warning here was something like 
```
except Exception e:
logger.warn("Error happened!")
```

The error log in this scenario is only serving to tell us that something bad happened / we hit this block of code, but doesn't give us any insight into what it was, which makes it harder for someone who sees this in the logs to go any further in debugging what happened. If something about the specific error was logged (e.g. the type of error, a stacktrace, information about the input parameters, etc. - depending on how you see the end user consuming/looking into the issue), it would help point the debugger in the right direction. 

Put another way, this log could say "SDNFallback: I'm here!" and it would serve the same purpose.

File: ecommerce/extensions/payment/views/cybersource.py (hunk 4 above)
Comment (dianekaplan) [reply ecommerce/extensions/payment/views/cybersource.py]: Updated to be more readable like this

File: ecommerce/extensions/payment/views/cybersource.py (hunk 11 above)
Comment (dianekaplan) [reply ecommerce/extensions/payment/views/cybersource.py]: thanks! Updated

File: ecommerce/extensions/payment/views/cybersource.py (hunk 12 above)
Comment (dianekaplan) [reply ecommerce/extensions/payment/views/cybersource.py]: Right- not sure if you saw my previous comment: I've updated to make sure we have specific logging for the known exception and any unexpected exception, so devs will have the info they need to investigate issues. In addition, I have this general log line- ("Skipped SDN API vs fallback comparison; exception calling checkSDNFallback")- happening in both cases, so we can search splunk once to get a sense of whether we have cases where we're unable to do the comparison.  Is there a better way to make this general issue query in splunk? Or was this comment just about the earlier catching exceptions with only this message, which kept the rest of the info from coming through? Okay to have the exception logged and also a more general log for splunk?

File: ecommerce/extensions/payment/core/sdn.py [hunk 13]
```diff
@@ -349,3 +350,52 @@ def populate_sdn_fallback_data_and_metadata(sdn_csv_string):
...
+    skip_message = "Skipped SDN API vs fallback comparison; exception calling checkSDNFallback"
+
+    try:
+        SDNFallback_hit = checkSDNFallback(
+            data['first_name'] + ' ' + data['last_name'],
+            data['city'],
+            data['country'])
+
+    # SDNFallbackDataEmptyError details logged previously in models code
+    except SDNFallbackDataEmptyError:
```
Comment (jhan217) [inline ecommerce/extensions/payment/core/sdn.py]: Thoughts on consolidating this with the general exception case to simplify the exception handling?

File: ecommerce/extensions/payment/core/sdn.py [hunk 14]
```diff
@@ -349,3 +350,52 @@ def populate_sdn_fallback_data_and_metadata(sdn_csv_string):
...
+        logger.warning(skip_message)
+    #  We're making this shadow call to gather comparison info vs SDN API. We want to know about any
+    # failures, but we don't want to gate/prevent purchases. Therefore we'll log the exception and pass
+    except Exception as e:  # pylint: disable=broad-except
```
Comment (jhan217) [inline ecommerce/extensions/payment/core/sdn.py]: Is this going to cover any exceptions thrown from code in the `else` block? Or are we 100% sure that nothing in the `else` block will ever throw an exception?

File: ecommerce/extensions/payment/core/sdn.py [hunk 15]
```diff
@@ -349,3 +350,52 @@ def populate_sdn_fallback_data_and_metadata(sdn_csv_string):
...
+        logger.exception(e)
+        logger.warning(skip_message)
+    else:
+        # If shadow fallback call ran, log comparison of results vs those from the SDN API
```
Comment (jhan217) [inline ecommerce/extensions/payment/core/sdn.py:380]: 🎱 minor: Is the code not clear enough to need this comment?

File: ecommerce/extensions/payment/core/tests/test_sdn.py [hunk 16]
```diff
@@ -445,3 +451,80 @@ def test_check_sdn_fallback_other_fields(self):
...
+                    self.LOGGER_NAME,
+                    'INFO',
+                    "SDNFallback compare: MATCH. Results - SDN API: 0 hit(s); SDN Fallback match: False. Basket: [999]"
+                )
+            )
+
+    def test_compare_SDNCheck_vs_fallback_mismatch(self):
+        """Log correct results from fallback and API calls: mismatch where API hit is missed in fallback"""
+
+        form_data_no_hit = {
```
Comment (jhan217) [inline ecommerce/extensions/payment/core/tests/test_sdn.py]: 🎱 minor: form_data_hit ?

File: ecommerce/extensions/payment/core/tests/test_sdn.py [hunk 17]
```diff
@@ -445,3 +451,80 @@ def test_check_sdn_fallback_other_fields(self):
...
+        with general warning. Trigger that exception by providing empty form data (second argument)
+        """
+
+        with LogCapture(self.LOGGER_NAME) as log_exception:
+            compare_SDNCheck_vs_fallback(999, {}, 1)
+            log_exception.check(
+                (
+                    self.LOGGER_NAME,
+                    'ERROR',
+                    "'first_name'"
```
Comment (jhan217) [inline ecommerce/extensions/payment/core/tests/test_sdn.py]: 🎱 Would it be useful to dump the parameters into the log in this case?

File: ecommerce/extensions/payment/tests/test_models.py [hunk 18]
```diff
@@ -310,3 +311,11 @@ def test_get_current_records_and_filter_by_source_and_type(self):
         filtered_records_isn = SDNFallbackData.get_current_records_and_filter_by_source_and_type(
             isn_source, "")
         self.assertEqual(len(filtered_records_isn), 1)
+
+    def test_get_current_records_and_filter_by_source_and_type_empty_data(self):
+        """ Verify that we raise the expected Exception if this is called before data is populated"""
+        sdn_source = "Specially Designated Nationals (SDN) - Treasury Department"
```
Comment (jhan217) [inline ecommerce/extensions/payment/tests/test_models.py:317]: 🎱 minor: not sure you need to put these into variables if they aren't getting used more than once

File: ecommerce/extensions/payment/core/sdn.py (hunk 15 above)
Comment (dianekaplan) [reply ecommerce/extensions/payment/core/sdn.py:380]: Emma had encouraged me to give plenty of context of why, since ordinarily we wouldn't do 'except Exception as e:'

File: ecommerce/extensions/payment/core/tests/test_sdn.py (hunk 16 above)
Comment (dianekaplan) [reply ecommerce/extensions/payment/core/tests/test_sdn.py]: The naming is from the SDN API results: that API returns a number of hits, so the existing sdn logging refers to if it got a hit or 0 hits. Would it be clearer as something like form_data_no_hit_from_sdn_api?

File: ecommerce/extensions/payment/core/sdn.py (hunk 15 above)
Comment (jhan217) [reply ecommerce/extensions/payment/core/sdn.py:380]: These comments don't reference exceptions? They seem to be just saying in plain English what the code below is doing

File: ecommerce/extensions/payment/core/tests/test_sdn.py (hunk 16 above)
Comment (jhan217) [reply ecommerce/extensions/payment/core/tests/test_sdn.py]: If I understand what's going on, `compare_SDNCheck_vs_fallback(basket_id, data, hit_count)` takes in basket_id, data from the form, and the number of matches from the API. Since the hit_count is 1 in the below code, does that mean that the data is associated with a hit, and therefore would be more accurately described as `form_data_hit`?

File: ecommerce/extensions/payment/core/sdn.py (hunk 13 above)
Comment (dianekaplan) [reply ecommerce/extensions/payment/core/sdn.py]: done! Emma and I talked about it more, and can differentiate the exceptions in the same line and have them do the same thing

File: ecommerce/extensions/payment/core/sdn.py (hunk 15 above)
Comment (dianekaplan) [reply ecommerce/extensions/payment/core/sdn.py:380]: Right- I think she wanted the plain english to explain that it's intentional that we're choosing to catch a general exception (usually not done) and this is why

File: ecommerce/extensions/payment/core/tests/test_sdn.py (hunk 16 above)
Comment (dianekaplan) [reply ecommerce/extensions/payment/core/tests/test_sdn.py]: Ah now I see why this is confusing- I used an SDN API shorthand to refer to the form data that's used for the fallback- "form_data_no_hit" is expected not to match the fallback, but in one test we'll pretend it got 0 hits from the API and in another we'll pretend it got 1 hit from the API. Sorry for the confusion! I'll rename this "form_data_not_in_fallback_csv"

File: ecommerce/extensions/payment/tests/test_models.py (hunk 18 above)
Comment (dianekaplan) [reply ecommerce/extensions/payment/tests/test_models.py:317]: True- but I like that it makes the method call shorter and more readable (for me)

File: ecommerce/extensions/payment/core/tests/test_sdn.py (hunk 17 above)
Comment (dianekaplan) [reply ecommerce/extensions/payment/core/tests/test_sdn.py]: (removed this test since it's no longer needed now that we've consolidated the exceptions)

File: ecommerce/extensions/payment/core/sdn.py (hunk 15 above)
Comment (dianekaplan) [reply ecommerce/extensions/payment/core/sdn.py:380]: (my mistake- this comment was about lines 379-380, not the earlier lines)- Emma thought this boolean had enough squinting that some english explanation would be helpful. Since it's an 8-ball suggestion I'll leave it as it is in case Emma's not the only one

File: ecommerce/extensions/payment/core/tests/test_sdn.py (hunk 16 above)
Comment (dianekaplan) [reply ecommerce/extensions/payment/core/tests/test_sdn.py]: Updated comments on these two tests to make it clearer what changes between them (just the hit_count passed), updated to pull the basket from the form_data, and renamed to 'form_data' to keep it simple. Will look into options for pulling the form_data out of those tests so they can share, but have it nearby for reference.

File: ecommerce/extensions/payment/core/sdn.py (hunk 14 above)
Comment (dianekaplan) [reply ecommerce/extensions/payment/core/sdn.py]: Discussed together with JJ and Emma- the code in the `else` block is not very deep or complex, and safer now that it's updated to use data.get (if data is missing it'll use null instead of throwing an exception). For that reason (and also because if someone's doing something sinister with the form data we don't WANT the purchase to go through) we'll leave that logging code exposed rather than putting into its own try/except block.

**Review by jhan217 — APPROVED** [review]

File: ecommerce/extensions/payment/views/cybersource.py [hunk 19]
```diff
@@ -84,6 +84,8 @@ def check_sdn(self, request, data):
             data['city'],
             data['country'])
 
+        compare_SDNCheck_vs_fallback(request.basket.id, data, hit_count)
```
Comment (jhan217) [inline ecommerce/extensions/payment/views/cybersource.py:87]: Just to close the loop on this - Per discussion with Emma + Diane, it was decided that there was low risk of this throwing an error and if it did throw an error, that we would want to (in the worst case) not complete the transaction and deal with it immediately.

File: ecommerce/extensions/payment/core/tests/test_sdn.py [hunk 20]
```diff
@@ -445,3 +451,63 @@ def test_check_sdn_fallback_other_fields(self):
...
+                    "SDNFallback compare: MATCH. Results - SDN API: 0 hit(s); SDN Fallback match: False. Basket: " +
+                    str(form_data['basket'])
+                )
+            )
+
+    def test_compare_SDNCheck_vs_fallback_mismatch(self):
+        """Log correct results from fallback and API calls: mismatch where API hit is missed in fallback
+        We'll use form data not matching fallback csv data, and pass 1 hit from the SDN API"""
+
+        form_data = {
```
Comment (julianajlk) [inline ecommerce/extensions/payment/core/tests/test_sdn.py:482]: Could this be placed in setUp to avoid duplicating the same form_data?

Comment (julianajlk) [reply ecommerce/extensions/payment/core/tests/test_sdn.py:482]: Nevermind, there are many other tests here, I was just looking at the changes instead of the entire file.

**Review by julianajlk — APPROVED** [review]

File: ecommerce/extensions/payment/core/tests/test_sdn.py (hunk 20 above)
Comment (dianekaplan) [reply ecommerce/extensions/payment/core/tests/test_sdn.py:482]: Good question! Yes I tried a few different ways of saving `form_data` outside of the tests to avoid the replication (including inside of setUp), but then it's 'not defined' down in the test. The other option would be to have one big test cover both cases, but that seems 'worse' than having that small block be repeated. In the end I think it's okay to leave, since this code and its tests are temporary.
